sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/ByteBuddyDoFnInvokerFactory.java

```diff
@@ -129,6 +130,32 @@ private ByteBuddyDoFnInvokerFactory() {}
     return newByteBuddyInvoker(DoFnSignatures.getSignature((Class) fn.getClass()), fn);
   }
 
+  /**
+   * Internal base class for generated {@link DoFnInvoker} instances.
+   *
+   * <p>This class should <i>not</i> be extended directly, or by Beam users. It must be public for
+   * generated instances to have adequate access, as they are generated "inside" the invoked {@link
+   * DoFn} class.
+   */
+  public abstract static class DoFnInvokerBase<InputT, OutputT, DoFnT extends DoFn<InputT, OutputT>>
+      implements DoFnInvoker<InputT, OutputT> {
+    protected DoFnT delegate;
+    private Map<String, OnTimerInvoker> onTimerInvokers = new HashMap<>();
+
+    public DoFnInvokerBase(DoFnT delegate) {
+      this.delegate = delegate;
+    }
+
+    void addOnTimerInvoker(String timerId, OnTimerInvoker onTimerInvoker) {
+      this.onTimerInvokers.put(timerId, onTimerInvoker);
+    }
+
+    @Override
+    public void invokeOnTimer(String timerId, DoFn.ExtraContextFactory<InputT, OutputT> args) {
+      onTimerInvokers.get(timerId).invokeOnTimer(args);
```

Comment:
I think we should probably do something better than throw an NPE when the timerId doesn't exist. I'd porpose:

```
OnTimerInvoker invoker = onTimerInvokers.get(timerId);
// This generally shouldn't happen. It would indicate that a timer firing has arrived for a timer that
// doesn't have a callback, but the reflection should validate that this is safe. One possible case
// where this could happen is after updating a pipeline that had a timer with a given ID set to no
// longer have that ID.
checkArgument(invoker != null,
  "Unrecognized timerId %s, expected one of %s", timerId, onTimerInvokers.keySet());
invoker.invokeOnTimer(args);
```

Also -- should the update case described be (1) silently dropped or (2) an update incompatible change?